### PR TITLE
GRPC

### PR DESCRIPTION
**Pull request by ajax-butriy-v**

### CODE REVIEW

File: gateway/src/main/kotlin/com/example/gateway/mapper/grpc/CreatePassMapper.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        }
+    }
+
+    private fun Pass.toCreatePassResponse(): CreatePassResponse {
+        return CreatePassResponse.newBuilder().also {
+            it.successBuilder.setPass(this)
+        }.build()
+    }
+
+    private fun InternalCreatePassResponse.failureCase(): Nothing {
```
Comment (ajax-torbin-v) [inline gateway/src/main/kotlin/com/example/gateway/mapper/grpc/CreatePassMapper.kt]:
```suggestion
    private fun InternalCreatePassResponse.Failure.asException(): Nothing {
```

File: gateway/build.gradle.kts [hunk 2]
```diff
@@ -9,6 +10,11 @@ dependencies {
     implementation(libs.spring.boot.starter.webflux)
     implementation(libs.spring.boot.starter.data.mongodb.reactive)
     implementation(libs.nats)
+    implementation("net.devh:grpc-spring-boot-starter:2.15.0.RELEASE")
```
Comment (ajax-malov-n) [inline gateway/build.gradle.kts]: Duplicated dependencies

File: gateway/src/main/kotlin/com/example/gateway/configuration/NatsClient.kt [hunk 3]
```diff
@@ -17,4 +22,15 @@ class NatsClient(private val natsConnection: Connection) {
...
     }
+
+    fun subscribeToPassesByType(passTypeName: String): Flux<Pass> {
+        val subjectName = NatsSubject.Pass.subjectByPassTypeName(passTypeName)
+        return Flux.create { fluxSink ->
+            val subscription = dispatcher.subscribe(subjectName) { message ->
+                val transferredMessage = TransferredPassMessage.parseFrom(message.data)
+                fluxSink.next(transferredMessage)
+            }
+            fluxSink.onDispose { dispatcher.unsubscribe(subscription) }
```
Comment (ajax-malov-n) [inline gateway/src/main/kotlin/com/example/gateway/configuration/NatsClient.kt]: Maybe add log here?

File: pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/kafka/configuration/KafkaReactiveProducerConfiguration.kt [hunk 4]
```diff
@@ -27,17 +27,15 @@ class KafkaReactiveProducerConfiguration(
...
-            .build()
+    fun topics(): NewTopics {
+        return NewTopics(
+            TopicBuilder.name(KafkaTopic.KafkaTransferPassEvents.TRANSFER)
+                .partitions(TRANSFER_PASS_TOPICS_PARTITIONS_AMOUNT)
+                .build(),
+            TopicBuilder.name(KafkaTopic.KafkaTransferPassEvents.TRANSFER_STATISTICS)
+                .partitions(TRANSFER_PASS_TOPICS_PARTITIONS_AMOUNT)
+                .build()
+        )
```
Comment (ajax-malov-n) [inline pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/kafka/configuration/KafkaReactiveProducerConfiguration.kt]: Divide into two beans with corresponding names

File: pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/kafka/consumer/TransferPassForNatsConsumer.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    private val transferPassForNatsKafkaReceiver: KafkaReceiver<String, ByteArray>,
+    private val passTypeService: PassTypeService,
+    private val natsConnection: Connection,
+) {
+    @EventListener(ApplicationReadyEvent::class)
+    fun consumeTransferMessageAndPublishToNats() {
+        transferPassForNatsKafkaReceiver.receiveBatch()
+            .flatMap { receiverRecords ->
+                receiverRecords.flatMap { record ->
+                    record.toMono()
```
Comment (ajax-malov-n) [inline pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/kafka/consumer/TransferPassForNatsConsumer.kt]: remove toMono()

File: pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/service/impl/PassManagementServiceImpl.kt [hunk 6]
```diff
@@ -28,11 +28,12 @@ internal class PassManagementServiceImpl(
             }
             .flatMap { (updatedPass, previousPassOwnerId) ->
                 val key = updatedPass.passTypeId.toString()
-                transferPassMessageProducer.sendTransferPassMessage(
+                val temp = transferPassMessageProducer.sendTransferPassMessage(
```
Comment (ajax-malov-n) [inline pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/service/impl/PassManagementServiceImpl.kt]: remove `val temp`

File: gateway/src/testFixtures/kotlin/com/example/gateway/util/PassGrpcProtoFixture.kt [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    fun successfulFindPassByIdResponse(pass: Pass): FindPassByIdResponse {
+        return FindPassByIdResponse.newBuilder().apply {
+            successBuilder.pass = pass
+        }.build()
+    }
+
+    fun failureFindPassByIdResponseWithPassNotFound(exceptionMessage: String): InternalFindPassByIdResponse {
+        return InternalFindPassByIdResponse.newBuilder().apply {
+            failureBuilder.notFoundById = Error.getDefaultInstance()
```
Comment (ajax-torbin-v) [inline gateway/src/testFixtures/kotlin/com/example/gateway/util/PassGrpcProtoFixture.kt]:
```suggestion
            failureBuilder.notFoundByIdBuilder
```

**Review by ajax-malov-n — COMMENTED** [review]
Ok

File: pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/mapper/PassOwnerMapper.kt [hunk 8]
```diff
@@ -1,8 +1,8 @@
-package com.example.passmanagersvc.web.mapper
+package com.example.passmanagersvc.mapper
 
 import com.example.passmanagersvc.domain.MongoPassOwner
-import com.example.passmanagersvc.web.dto.PassOwnerDto
-import com.example.passmanagersvc.web.dto.PassOwnerUpdateDto
+import com.example.passmanagersvc.dto.PassOwnerDto
+import com.example.passmanagersvc.dto.PassOwnerUpdateDto
 
 internal object PassOwnerMapper {
```
Comment (ajax-torbin-v) [inline pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/mapper/PassOwnerMapper.kt]: Internal?

File: common-proto/src/main/proto/com/example/commonmodel/pass_owner.proto [hunk 9]
```diff
@@ -1,6 +1,6 @@
 syntax = "proto3";
 
-package com.example.internal.commonmodels;
+package com.example.commonmodel;
```
Comment (ajax-shulepa-a) [inline common-proto/src/main/proto/com/example/commonmodel/pass_owner.proto]:
```suggestion
package com.example.commonmodels;
```

File: gateway/src/main/kotlin/com/example/gateway/configuration/NatsClient.kt [hunk 10]
```diff
@@ -17,4 +23,22 @@ class NatsClient(private val natsConnection: Connection) {
...
+            }
+            fluxSink.onDispose {
+                log.info("NATS Finalizing subscription to $passTypeName")
+                dispatcher.unsubscribe(subscription)
+            }
+        }.map(TransferredPassMessage::getPass)
+    }
+
+    companion object {
+        private val log = LoggerFactory.getLogger(this::class.java)
```
Comment (ajax-shulepa-a) [inline gateway/src/main/kotlin/com/example/gateway/configuration/NatsClient.kt]: what will be the class name in the logger?

File: gateway/src/main/kotlin/com/example/gateway/grpc/PassGrpcService.kt [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+import com.example.internal.NatsSubject.Pass.CREATE
+import com.example.internal.NatsSubject.Pass.FIND_BY_ID
+import net.devh.boot.grpc.server.service.GrpcService
+import reactor.core.publisher.Flux
+import reactor.core.publisher.Mono
+import com.example.internal.input.reqreply.CreatePassResponse as InternalCreatePassResponse
+import com.example.internal.input.reqreply.FindPassByIdResponse as InternalFindPassByIdResponse
+
+@GrpcService
+class PassGrpcService(
```
Comment (ajax-shulepa-a) [inline gateway/src/main/kotlin/com/example/gateway/grpc/PassGrpcService.kt]:
```suggestion
class GrpcPassService(
```

File: grpc-api/src/main/proto/com/example/grpcapi/reqrep/pass/create_pass.proto [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+package com.example.grpcapi.reqrep.pass;
+
+import "com/example/commonmodel/pass.proto";
+import "com/example/commonmodel/decimal.proto";
+import "com/example/commonmodel/error.proto";
+
+option java_multiple_files = true;
+
+message CreatePassRequest {
+  commonmodel.Pass passToCreate = 1;
```
Comment (ajax-shulepa-a) [inline grpc-api/src/main/proto/com/example/grpcapi/reqrep/pass/create_pass.proto]:
```suggestion
  commonmodel.Pass pass_to_create = 1;
```

**Review by ajax-shulepa-a — APPROVED** [review]

**Pull request merged**